Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm

```diff
@@ -0,0 +1,264 @@
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ *
+ * THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS ``AS IS''
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ * THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ * PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS
+ * BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+ * THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+#if !__has_feature(objc_arc)
+#error This file requires ARC. Add the "-fobjc-arc" compiler flag for this file.
+#endif
+
+#if ENABLE(WK_WEB_EXTENSIONS)
+
+#import "config.h"
+#import "WebExtensionAPIScripting.h"
+
+#import "CocoaHelpers.h"
+#import "Logging.h"
+#import "MessageSenderInlines.h"
+#import "WebExtension.h"
+#import "WebExtensionAPINamespace.h"
+#import "WebExtensionContextMessages.h"
+#import "WebExtensionUtilities.h"
+#import "WebProcess.h"
+
+static NSString * const allFramesKey = @"allFrames";
+static NSString * const argsKey = @"args";
+static NSString * const argumentsKey = @"arguments";
+static NSString * const filesKey = @"files";
+static NSString * const frameIDsKey = @"frameIds";
+static NSString * const funcKey = @"func";
+static NSString * const functionKey = @"function";
+static NSString * const tabIDKey = @"tabId";
+static NSString * const targetKey = @"target";
+static NSString * const worldKey = @"world";
+
+static NSString * const  cssKey = @"css";
+// FIXME: <https://webkit.org/b/261765> Consider adding support for cssOrigin.
+
+static NSString * const mainWorld = @"MAIN";
+static NSString * const isolatedWorld = @"ISOLATED";
+
+static int64_t const mainFrameID = 0;
+
+// FIXME: <https://webkit.org/b/261765> Consider adding support for injectImmediately.
+
+namespace WebKit {
+
+void WebExtensionAPIScripting::executeScript(NSDictionary *script, Ref<WebExtensionCallbackHandler>&& callback, NSString **outExceptionString)
+{
+    // Documentation: https://developer.mozilla.org/en-US/docs/Mozilla/Add-ons/WebExtensions/API/scripting/executeScript
+
+    if (!validateScript(script, outExceptionString))
+        return;
+
+    // A JSValue cannot be transferred across processes, so we must convert it to a string before passing it along.
+    if (JSValue *func = script[funcKey] ?: script[functionKey]) {
+        ASSERT(func.isObject);
```

Comment:
This should be more than an ASSERT, since validate only checks NSObject.class currently. This should be an exception if `!func.isObject`.

Reply:
Spoke to @xeenon about this offline. Doesn't look like there is a way for anything to pass the `JSValue.class` check but fail `isObject`. Will change the validation to check against `JSValue` instead of `NSObject` but will remove this `isObject` check.